File: android/app/src/main/kotlin/com/app/deenly/deenly/qurantranslation/QuranTranslationChannelHandler.kt
package com.rnr.deenfocus.qurantranslation

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File

/** MethodChannel + EventChannel for downloadable Quran translations (ADR-009). */
class QuranTranslationChannelHandler(context: Context) : EventChannel.StreamHandler {
    private val appContext = context.applicationContext
    private val mainHandler = Handler(Looper.getMainLooper())
    private var eventSink: EventChannel.EventSink? = null

    private val store get() = TranslationStore.getInstance(appContext)
    private val catalogCache get() = TranslationCatalogCache(File(appContext.filesDir, "QuranTranslations"))

    fun attach(methodChannel: MethodChannel, eventChannel: EventChannel) {
        eventChannel.setStreamHandler(this)
        methodChannel.setMethodCallHandler { call, result -> handle(call, result) }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    private fun emitProgress(progress: Double) {
        mainHandler.post {
            eventSink?.success(
                mapOf(
                    "type" to "downloadProgress",
                    "progress" to progress,
                ),
            )
        }
    }

    private fun handle(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "listAvailableTranslations" -> runAsync(result) {
                val forceRefresh = call.argument<Boolean>("forceRefresh") ?: true
                val entries = if (forceRefresh) {
                    try {
                        catalogCache.translationEntries(forceRefresh = true)
                    } catch (_: Exception) {
                        // Offline / transient — show last known catalog rather than empty.
                        try {
                            catalogCache.translationEntries(forceRefresh = false)
                        } catch (_: Exception) {
                            emptyList()
                        }
                    }
                } else {
                    catalogCache.translationEntries(forceRefresh = false)
                }
                entries.map { entry ->
                    buildMap<String, Any?> {
                        put("language", entry.language)
                        put("packId", entry.packId)
                        put("displayName", entry.displayName)
                        put("installed", store.isAvailable(entry.language))
                        entry.approxSizeBytes?.let { put("approxSizeBytes", it) }
                    }
                }
            }
            "hasTranslationPack" -> {
                val languageCode = languageArg(call) ?: run {
                    result.error("INVALID_ARGS", "languageCode required", null)
                    return
                }
                runAsync(result) { catalogCache.entryForLanguage(languageCode) != null }
            }
            "isTranslationAvailable" -> {
                val languageCode = languageArg(call) ?: run {
                    result.error("INVALID_ARGS", "languageCode required", null)
                    return
                }
                result.success(store.isAvailable(languageCode))
            }
            "getTranslationPath" -> {
                val languageCode = languageArg(call) ?: run {
                    result.error("INVALID_ARGS", "languageCode required", null)
                    return
                }
                val file = store.translationFile(languageCode)
                result.success(if (file.exists()) file.absolutePath else null)
            }
            "ensureTranslation" -> {
                val languageCode = languageArg(call) ?: run {
                    result.error("INVALID_ARGS", "languageCode required", null)
                    return
                }
                runAsync(result) {
                    store.ensureTranslation(languageCode) { progress -> emitProgress(progress) }
                    true
                }
            }
            "refreshCatalog" -> runAsync(result) {
                catalogCache.catalog(forceRefresh = true)
                true
            }
            else -> result.notImplemented()
        }
    }

    private fun languageArg(call: MethodCall): String? =
        call.argument<String>("languageCode")
            ?: (call.arguments as? Map<*, *>)?.get("languageCode") as? String

    private fun <T> runAsync(result: MethodChannel.Result, block: () -> T) {
        Thread {
            try {
                val value = block()
                mainHandler.post { result.success(value) }
            } catch (e: Exception) {
                mainHandler.post { result.error("DOWNLOAD_FAILED", e.message, null) }
            }
        }.start()
    }
}

File: ios/Runner/QuranTranslation/QuranTranslationChannelHandler.swift
import Flutter
import Foundation

/// MethodChannel + EventChannel for downloadable Quran translations (ADR-009).
final class QuranTranslationChannelHandler: NSObject, FlutterStreamHandler {
  private var eventSink: FlutterEventSink?
  private let workQueue = DispatchQueue(label: "com.app.deenly.quran_translation", qos: .utility)

  func register(messenger: FlutterBinaryMessenger) {
    let method = FlutterMethodChannel(
      name: "com.app.deenly.deenly/quran_translations",
      binaryMessenger: messenger
    )
    let events = FlutterEventChannel(
      name: "com.app.deenly.deenly/quran_translation_events",
      binaryMessenger: messenger
    )
    events.setStreamHandler(self)
    method.setMethodCallHandler { [weak self] call, result in
      self?.handle(call: call, result: result)
    }
  }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  private func emitProgress(_ progress: Double) {
    DispatchQueue.main.async { [weak self] in
      self?.eventSink?([
        "type": "downloadProgress",
        "progress": progress,
      ])
    }
  }

  private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "listAvailableTranslations":
      workQueue.async {
        do {
          let args = call.arguments as? [String: Any]
          let forceRefresh = (args?["forceRefresh"] as? Bool) ?? true
          let entries: [TranslationCatalogEntry]
          if forceRefresh {
            do {
              entries = try TranslationCatalogCache.shared.translationEntries(forceRefresh: true)
            } catch {
              // Offline / transient — show last known catalog rather than empty.
              entries = (try? TranslationCatalogCache.shared.translationEntries(forceRefresh: false)) ?? []
            }
          } else {
            entries = try TranslationCatalogCache.shared.translationEntries(forceRefresh: false)
          }
          let payload = entries.map { entry -> [String: Any] in
            var row: [String: Any] = [
              "language": entry.language,
              "packId": entry.packId,
              "displayName": entry.displayName as Any,
              "installed": TranslationStore.shared.isAvailable(entry.language),
            ]
            if let size = entry.approxSizeBytes {
              row["approxSizeBytes"] = size
            }
            return row
          }
          DispatchQueue.main.async { result(payload) }
        } catch {
          DispatchQueue.main.async {
            result(FlutterError(code: "CATALOG_FAILED", message: error.localizedDescription, details: nil))
          }
        }
      }

    case "hasTranslationPack":
      guard let args = call.arguments as? [String: Any],
        let languageCode = args["languageCode"] as? String
      else {
        result(FlutterError(code: "INVALID_ARGS", message: "languageCode required", details: nil))
        return
      }
      workQueue.async {
        let has = (try? TranslationCatalogCache.shared.entry(forLanguage: languageCode)) != nil
        DispatchQueue.main.async { result(has) }
      }

    case "isTranslationAvailable":
      guard let args = call.arguments as? [String: Any],
        let languageCode = args["languageCode"] as? String
      else {
        result(FlutterError(code: "INVALID_ARGS", message: "languageCode required", details: nil))
        return
      }
      result(TranslationStore.shared.isAvailable(languageCode))

    case "getTranslationPath":
      guard let args = call.arguments as? [String: Any],
        let languageCode = args["languageCode"] as? String
      else {
        result(FlutterError(code: "INVALID_ARGS", message: "languageCode required", details: nil))
        return
      }
      let path = TranslationStore.shared.translationFileURL(languageCode)
      result(FileManager.default.fileExists(atPath: path.path) ? path.path : NSNull())

    case "ensureTranslation":
      guard let args = call.arguments as? [String: Any],
        let languageCode = args["languageCode"] as? String
      else {
        result(FlutterError(code: "INVALID_ARGS", message: "languageCode required", details: nil))
        return
      }
      workQueue.async { [weak self] in
        do {
          try TranslationStore.shared.ensureTranslation(languageCode: languageCode) { progress in
            self?.emitProgress(progress)
          }
          DispatchQueue.main.async { result(true) }
        } catch {
          DispatchQueue.main.async {
            result(FlutterError(code: "DOWNLOAD_FAILED", message: error.localizedDescription, details: nil))
          }
        }
      }

    case "refreshCatalog":
      workQueue.async {
        do {
          _ = try TranslationCatalogCache.shared.catalog(forceRefresh: true)
          DispatchQueue.main.async { result(true) }
        } catch {
          DispatchQueue.main.async {
            result(FlutterError(code: "CATALOG_FAILED", message: error.localizedDescription, details: nil))
          }
        }
      }

    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
